src/java.base/share/classes/jdk/internal/ffi/util/FFMUtils.java

```diff
@@ -0,0 +1,131 @@
+    public static final AddressLayout C_POINTER = ValueLayout.ADDRESS
+            .withTargetLayout(MemoryLayout.sequenceLayout(Long.MAX_VALUE, JAVA_BYTE));
+    public static final ValueLayout.OfLong C_LONG = (ValueLayout.OfLong) Linker.nativeLinker().canonicalLayouts().get("long");
+    private FFMUtils() {
+    }
+
+    /**
+     * Returns a {@code MemorySegment} set to the size of byteSize
+     *
+     * @param byteSize the size in bytes to be allocated
+     * @param byteAlignment the size in bytes for the memory alignment
+     *
+     * @throws IllegalArgumentException if the maxByteAlignment of the created
+     * MemorySegment is less than the provided byteAlignment
+     *
+     * @return the newly created {@code MemorySegment}
+     */
+    public static MemorySegment malloc(long byteSize, long byteAlignment) {
+        long allocatedMemory = UNSAFE.allocateMemory(byteSize);
+        MemorySegment result = MemorySegment.ofAddress(allocatedMemory).reinterpret(byteSize);
+        if (result.maxByteAlignment() < byteAlignment) {
+            throw new IllegalArgumentException();
+        }
+        return result;
+    }
+
+    /**
+     * Takes a {@code MemorySegment} and deallocates the memory at that address
+     * @param memorySegment the {@code MemorySegment} that will be deallocated
+     */
+    public static void free(MemorySegment memorySegment) {
+        UNSAFE.freeMemory(memorySegment.address());
+    }
+
+    // SegmentAllocator that delegates to Unsafe for memory allocation
+    public static final SegmentAllocator SEGMENT_ALLOCATOR = new SegmentAllocator() {
+        @Override
+        public MemorySegment allocate(long byteSize, long byteAlignment) {
+            return malloc(byteSize, byteAlignment);
+        }
+    };
+
+    private static final Unsafe UNSAFE = Unsafe.getUnsafe();
+
+    // Variables and methods below are extracted from jextract generated
+    // code and used by native bindings on all platforms
+    public static final boolean TRACE_DOWNCALLS = false;
+    static final SymbolLookup SYMBOL_LOOKUP = SymbolLookup.loaderLookup()
+            .or(Linker.nativeLinker().defaultLookup());
+
+    public static void traceDowncall(String name, Object... args) {
+        String traceArgs = Arrays.stream(args)
+                .map(Object::toString)
+                .collect(Collectors.joining(", "));
+        System.out.printf("%s(%s)\n", name, traceArgs);
+    }
+
+    public static MemorySegment findOrThrow(String symbol) {
+        return SYMBOL_LOOKUP.findOrThrow(symbol);
+    }
+
+    public static MethodHandle upcallHandle(Class<?> fi, String name, FunctionDescriptor fdesc) {
+        try {
+            return MethodHandles.lookup().findVirtual(fi, name, fdesc.toMethodType());
+        } catch (ReflectiveOperationException ex) {
+            throw new AssertionError(ex);
+        }
+    }
+
+    public static MemoryLayout align(MemoryLayout layout, long align) {
```

Comment:
As observed in the past, this `align` method could probably just be added on `MemoryLayout` (not as part of this PR, this is just a general observation).